app/src/main/java/org/odk/odknotifications/Activities/MainActivity.java

```diff
@@ -293,9 +254,51 @@ public void onBackPressed() {
         }
     }
 
+    private void addNotifications(){
+
+        RecyclerView recyclerView = findViewById(R.id.rv_notifications);
+        notificationArrayList = dbHandler.getAllNotificationsWithResponses();
+        notificationAdapter = new NotificationAdapter(notificationArrayList, this);
+        TextView textView = findViewById(R.id.no_data_text_view);
+
+        if(notificationArrayList == null || notificationArrayList.size()==0) {
```

Comment:
This handling needs to be done whenever this list is updated not just at the time of onCreate. And there are still some places where it will not work. This thing should only handle the logic of creating the adapter of notification list, nothing else. Rest part like changing textview visibility or text should be done separately. You can create separate method which would basically handle the post list creation part.

```
private void changeUIVisibilityWhenAdapterUpdates() {
    if(notificationArrayList == null || notificationArrayList.size()==0) {
            textView.setText(R.string.No_data);
            textView.setVisibility(View.VISIBLE);
     } else {
            textView.setVisibility(View.GONE);
     }
}
```

And addnotification method should like below

```
private void addNotification() {
        notificationArrayList = dbHandler.getAllNotificationsWithResponses();
        notificationAdapter = new NotificationAdapter(notificationArrayList, this);
        TextView textView = findViewById(R.id.no_data_text_view);
         notificationAdapter.setonButtonClickListener(new NotificationAdapter.onButtonClickListener() {
                @Override
                public boolean onSendResponseButtonClick(int position, String response) {
                    Notification notification = notificationArrayList.get(position);
                    ResponseHandler responseHandler = new ResponseHandler(getApplicationContext());
                    boolean isDone = responseHandler.saveResponse(notification.getId(), response, new Date().getTime());
                    if (isDone) {
                        Toast.makeText(getApplicationContext(), "Message sent successfully", Toast.LENGTH_LONG).show();
                        return true;
                    } else {
                        Toast.makeText(getApplicationContext(), "Some error occurred. Please try again later", Toast.LENGTH_LONG).show();
                        return false;
                    }
                }
            });
}
```

And assign the `textView` and `recyclerView` in onCreate itself. Try to avoid assigning the UI elements in methods because when more and more people start contributing methods evolve and context of them changes with more and more features, so assignment should ideally be handled by the method caller.

Reply:
Done.